arangod/Replication2/ReplicatedLog/Components/StorageManager.cpp

```diff
@@ -97,10 +97,58 @@ struct comp::StorageManagerTransaction : IStorageTransaction {
     mapping.append(sliceMapping);
     return scheduleOperation(
         std::move(mapping),
-        [slice = std::move(slice), iter = std::move(iter)](
+        [slice = std::move(slice), iter = std::move(iter),
+         weakManager = manager.weak_from_this(), waitForSync](
             StorageManager::IStorageEngineMethods& methods) mutable noexcept {
-          return methods.insert(std::move(iter), {.waitForSync = true})
-              .thenValue([](auto&& res) { return res.result(); });
+          auto lastIndex = slice.getLastIndex();
+          auto&& fut =
+              methods.insert(std::move(iter), {.waitForSync = waitForSync});
+
+          if (waitForSync) {
+            return std::move(fut).thenValue(
+                [lastIndex, weakManager = std::move(weakManager)](auto&& res) {
+                  if (auto mngr = weakManager.lock(); res.ok() && mngr) {
+                    mngr->syncIndex.doUnderLock([lastIndex](auto&& syncIndex) {
+                      if (syncIndex < lastIndex) {
+                        syncIndex = lastIndex;
+                      }
+                    });
+                  }
+                  return res.result();
+                });
+          }
+
+          return std::move(fut).thenValue([lastIndex,
+                                           weakManager = std::move(weakManager),
+                                           &methods](auto&& res) mutable {
+            // We're done writing to the WAL. That doesn't necessarily
+            // mean the data is synced to disk.
+            if (auto mngr = weakManager.lock(); res.ok() && mngr) {
+              // Methods are available as long as the manager is not
+              // destroyed.
+              mngr->scheduler->queue(
+                  [lastIndex, weakManager = std::move(weakManager),
+                   wfs = methods.waitForSync(res.get())]() mutable {
+                    auto syncRes = wfs.get();
+                    if (auto mngr = weakManager.lock()) {
```

Comment:
What's this stunt supposed to do? We create a future with `waitForSync` and then put in a lambda, post that on the scheduler and then synchronously wait for the future in a new thread. Is there a problem with `thenValue`?

Reply:
The new version is using `thenFinal`.